Competitive Programming/c2cpp.md

```diff
@@ -0,0 +1,117 @@
+# C to Cpp for competitive
+
+This article is written for specifically keeping competetive in mind and not for using cpp in development or any other context. Also this aims at getting started to use good and important features of c++ as quickly as possible so skips many details.
+Most of the c++ fetures talked here work on c++98 and after but some work only after c++11 so for simplicity we will assume to be using c++11. Also we will be using gcc/g++ compiler. You can also use an IDE like codeblocks it already has gcc and g++ compilers by default (MINGW); remember to enable c++11 or c++14 option in codebloks' settings.
+
+Once you are familiar with C its easy to migrate to and learn C++. First of all since c++ is superset of C so you can use most of the concepts of c in c++. So we will first talk about porting code and then a couple of things new in c++ useful for competetive.
+
+# PREREQUISITES
+We require that you are atleast aware of the following concepts in c:
+
+* Basic Syntax of C (input, output, and logical, relational, arithmetic & bitwise operators):  
+* Loops
+* Arrays/Strings
+* Functions
+* Structures
+* Pointers
+
+If you don't know any of the above things, its better to have a recap first and than come back to this article.
+Also we WON'T require things like enums, unions, static, file IO etc. so as long as you are familiar with the ones above you are good to go.
+
+# PORTING C CODE
+So all of the things listed in prerequisites remain same in C++. You can use the same `scanf` and `printf` in C++ also; they are defined in header files <stdio.h> and <cstdio>. Similarly string functions are defined in header files <string.h> and <cstring>, math functions in math.h and <cmath>. Though the files listed in pairs are same for most intend and purposes but have a subtle difference which requires knowledge of namespaces. Though we will not go into this discussion but I have highlighted a few recources at the end of this section for any qurious mind. Though NOT recommended for beginners.
+
+For competitive programmers in general tend to use `#include<bits/stdc++.h>` which is basically a header file containing all the standard c++ header files (most STL files, header files for c etc.), though again its not a good developement practice to include unnecesarry headers but it is pretty common in competitive, as judges use runtime for time limit and this only increases my compile time and not my runtime. For more details refer to [geekforgeeks](http://www.geeksforgeeks.org/bitsstdc-h-c/)
```

Comment:
I think some judges include compile time in run-time, since otherwise people can use [compile-time execution](https://en.wikipedia.org/wiki/Compile_time_function_execution) using template metaprogramming. Also, see this - http://codeforces.com/blog/entry/17026.

Reply:
I looked at the second link, it seems like the TLE was due to fast IO. I would like to try compile time execution first for the exact picture. Do you know any specific problem on which we can experiment this easily??

Reply:
After a more careful look at Akul's blog post, it seems more likely that the problem arose because of changes in runtime environment and not because of increase in compile time. I think the only way to test this is to craft a program which has an enormous compilation time (perhaps using template-metaprogramming to solve an exponential-time problem) but low running time.

Reply:
Some people, like I, don't like the increase in local compile time caused by `<bits/stdc++.h>`. So I think you should suggest the alternative of including header files in a template.